doc/libpmempool.3.md
---
layout: manual
Content-Style: 'text/css'
title: _MP(LIBPMEMPOOL, 3)
header: NVM Library
date: pmempool API version 1.1
...

[comment]: <> (Copyright 2016-2017, Intel Corporation)

[comment]: <> (Redistribution and use in source and binary forms, with or without)
[comment]: <> (modification, are permitted provided that the following conditions)
[comment]: <> (are met:)
[comment]: <> (    * Redistributions of source code must retain the above copyright)
[comment]: <> (      notice, this list of conditions and the following disclaimer.)
[comment]: <> (    * Redistributions in binary form must reproduce the above copyright)
[comment]: <> (      notice, this list of conditions and the following disclaimer in)
[comment]: <> (      the documentation and/or other materials provided with the)
[comment]: <> (      distribution.)
[comment]: <> (    * Neither the name of the copyright holder nor the names of its)
[comment]: <> (      contributors may be used to endorse or promote products derived)
[comment]: <> (      from this software without specific prior written permission.)

[comment]: <> (THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS)
[comment]: <> ("AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT)
[comment]: <> (LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR)
[comment]: <> (A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT)
[comment]: <> (OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,)
[comment]: <> (SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT)
[comment]: <> (LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,)
[comment]: <> (DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY)
[comment]: <> (THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT)
[comment]: <> ((INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE)
[comment]: <> (OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.)

[comment]: <> (libpmempool.3 -- man page for libpmempool)

[NAME](#name)<br />
[SYNOPSIS](#synopsis)<br />
[DESCRIPTION](#description)<br />
[POOL CHECKING FUNCTIONS](#pool-checking-functions)<br />
[POOL SET SYNCHRONIZATION AND TRANSFORMATION](#pool-set-synchronization-and-transformation-1)<br />
[POOL SET MANAGEMENT FUNCTIONS](#pool-set-management-functions)<br />
[LIBRARY API VERSIONING](#library-api-versioning-1)<br />
[DEBUGGING AND ERROR HANDLING](#debugging-and-error-handling)<br />
[EXAMPLE](#example)<br />
[ACKNOWLEDGEMENTS](#acknowledgements)<br />
[SEE ALSO](#see-also)<br />


# NAME #

**libpmempool** -- pool management library


# SYNOPSIS #

```c
#include <libpmempool.h>
cc -std=gnu99 ... -lpmempool -lpmem
```

_WINUX(
_q_>NOTE: NVML API supports UNICODE. If **NVML_UTF8_API** macro is defined then
basic API functions are expanded to UTF-8 API with postfix *U*,
otherwise they are expanded to UNICODE API with postfix *W*._e_)

##### Health check functions: #####

```c
_UWFUNCR1UW(PMEMpoolcheck, *pmempool_check_init, struct pmempool_check_args, *args,_q_
	size_t args_size_e_)
_UWFUNCRUW(struct pmempool_check_status, *pmempool_check, PMEMpoolcheck *ppc)
enum pmempool_check_result pmempool_check_end(PMEMpoolcheck *ppc);
```

##### Pool set synchronization and transformation: #####

```c
_UWFUNCR1(int, pmempool_sync, *poolset_file,_q_
	unsigned flags_e_, _q_ (EXPERIMENTAL)_e_)
_UWFUNCR12(int, pmempool_transform, *poolset_file_src,
	*poolset_file_dst, unsigned flags, _q_ (EXPERIMENTAL)_e_)
```

##### Pool set management functions: #####

```c
_UWFUNCR1(int, pmempool_rm, *path, int flags)
```

##### Library API versioning: #####

```c
_UWFUNC(pmempool_check_version, _q_
	unsigned major_required,
	unsigned minor_required_e_)
```

##### Error handling: #####

```c
_UWFUNC(pmempool_errormsg, void)
```


# DESCRIPTION #

**libpmempool**
provides a set of utilities for off-line analysis and
manipulation of a *pool*. By *pool* in this
manpage we mean pmemobj pool, pmemblk pool, pmemlog pool or
BTT layout, independent of the underlying storage. Some
**libpmempool** functions are required to work without
any impact on the *pool* but some may create a new or modify
an existing *pool*.

**libpmempool**
is for applications that need high reliability or built-in
troubleshooting. It may be useful for testing and debugging
purposes also.


# POOL CHECKING FUNCTIONS #

To perform the checks provided by **libpmempool**, a *check context*
must first be initialized using the _UW(pmempool_check_init)
function described in this section. Once initialized, the
*check context* is represented by an opaque handle of
type *PMEMpoolcheck\**, which is passed to all of the
other functions described in this section.

To execute checks, _UW(pmempool_check) must be called iteratively.
Each call generates a new check status, represented by a
_UWS(pmempool_check_status) structure. Status messages are described
later in this section.

When the checks are completed, _UW(pmempool_check) returns NULL.
The check must be finalized using **pmempool_check_end**(), which
returns an *enum pmempool_check_result* describing the results of
the entire check.

> NOTE: Currently, checking the consistency of a *pmemobj* pool is
**not** supported.

```c
_UWFUNCR1UW(PMEMpoolcheck, *pmempool_check_init, struct pmempool_check_args,
	*args,_q_
	size_t args_size_e_)
```

_UW(pmempool_check_init) initializes the check
context. *args* describes parameters of the
check context. *args_size* should be equal to
the size of the _UWS(pmempool_check_args).
_UWS(pmempool_check_args) is defined as follows:

_WINUX(_q_
```c
struct pmempool_check_argsU
{
	/* path to the pool to check */
	const char *path;

	/* optional backup path */
	const char *backup_path;

	/* type of the pool */
	enum pmempool_pool_type pool_type;

	/* parameters */
	int flags;
};

struct pmempool_check_argsW
{
	/* path to the pool to check */
	const wchar_t *path;

	/* optional backup path */
	const wchar_t *backup_path;

	/* type of the pool */
	enum pmempool_pool_type pool_type;

	/* parameters */
	int flags;
};
```
_e_,_q_
```c
struct pmempool_check_args
{
	/* path to the pool to check */
	const char *path;

	/* optional backup path */
	const char *backup_path;

	/* type of the pool */
	enum pmempool_pool_type pool_type;

	/* parameters */
	int flags;
};
```
_e_)
The *flags* argument accepts any combination of the following values (ORed):

+ **PMEMPOOL_CHECK_REPAIR** - perform repairs
+ **PMEMPOOL_CHECK_DRY_RUN** - emulate repairs, not supported on Device DAX
+ **PMEMPOOL_CHECK_ADVANCED** - perform hazardous repairs
+ **PMEMPOOL_CHECK_ALWAYS_YES** - do not ask before repairs
+ **PMEMPOOL_CHECK_VERBOSE** - generate info statuses
+ **PMEMPOOL_CHECK_FORMAT_STR** - generate string format statuses

If any of the provided parameters are invalid, or the initialization process
fails, _UW(pmempool_check_init) returns NULL and sets *errno*
appropriately. *pool_type* must match the type of the
*pool* being processed. You can turn on pool type
detection by setting *pool_type* to **PMEMPOOL_POOL_TYPE_DETECT**.
A pool type detection failure ends the check.

*backup_path* may be:

+ NULL: No backup will be performed.

+ a non-existent file: *backup_path* will be created and backup will be performed. *path* must be a single file *pool*.

+ an existing pool set file: Backup will be performed
as defined by the *backup_path* pool set. *backup_path* must have the same structure (the same number of parts with exactly the same size) as the
*path* pool set.

Backup is supported only if the source pool set has no defined replicas.

Neither *path* nor *backup_path* may specify a pool set with remote replicas.

This is an example of a *check context* initialization:

```c
struct _U(pmempool_check_args) args =
{
	.path = "/path/to/blk.pool",
	.backup_path = NULL,
	.pool_type = PMEMPOOL_POOL_TYPE_BLK,
	.flags = PMEMPOOL_CHECK_REPAIR | PMEMPOOL_CHECK_DRY_RUN |
		PMEMPOOL_CHECK_VERBOSE | PMEMPOOL_CHECK_FORMAT_STR
};
```

```c
PMEMpoolcheck *ppc = _U(pmempool_check_init)(&args, sizeof(args));
```

The check will process a *pool* of type **PMEMPOOL_POOL_TYPE_BLK**
located in the path */path/to/blk.pool*. Before check it will
not create a backup of the *pool* (*backup_path == NULL*).
If the check find anys issues it will try to
perform repair steps (**PMEMPOOL_CHECK_REPAIR**), but it
will not make any changes to the *pool*
(**PMEMPOOL_CHECK_DRY_RUN**) and it will not perform any
dangerous repair steps (no **PMEMPOOL_CHECK_ADVANCED**).
The check will ask before performing any repair steps (no
**PMEMPOOL_CHECK_ALWAYS_YES**). It will also generate
detailed information about the check (**PMEMPOOL_CHECK_VERBOSE**).
**PMEMPOOL_CHECK_FORMAT_STR** flag indicates string
format statuses (*struct pmempool_check_status*).
Currently this is the only supported status format so this flag is required.

```c
_UWFUNCRUW(struct pmempool_check_status, *pmempool_check, PMEMpoolcheck *ppc)
```

The _UW(pmempool_check) function starts or resumes the check
indicated by *ppc*. When next status will be generated
it pauses the check and returns a pointer to the
_UWS(pmempool_check_status) structure:

_WINUX(
_q_
```c
struct pmempool_check_statusU
{
	enum pmempool_check_msg_type type; /* type of the status */
	struct
	{
		const char *msg; /* status message string */
		const char *answer; /* answer to message if applicable */
	} str;
};

struct pmempool_check_statusW
{
	enum pmempool_check_msg_type type; /* type of the status */
	struct
	{
		const wchar_t *msg; /* status message string */
		const wchar_t *answer; /* answer to message if applicable */
	} str;
};
```
_e_,_q_
```c
struct pmempool_check_status
{
	enum pmempool_check_msg_type type; /* type of the status */
	struct
	{
		const char *msg; /* status message string */
		const char *answer; /* answer to message if applicable */
	} str;
};
```
_e_)

This structure can describe three types of statuses:

+ **PMEMPOOL_CHECK_MSG_TYPE_INFO** - detailed information about the check.
  Generated only if a **PMEMPOOL_CHECK_VERBOSE** flag was set.
+ **PMEMPOOL_CHECK_MSG_TYPE_ERROR** - encountered error
+ **PMEMPOOL_CHECK_MSG_TYPE_QUESTION** - question. Generated only if the
  **PMEMPOOL_CHECK_ALWAYS_YES** flag was not set. It requires *answer* to be
  set to "yes" or "no" before continuing.

After calling _UW(pmempool_check) again, the previously provided
_UWS(pmempool_check_status) pointer must be
considered invalid. When the check completes,
_UW(pmempool_check) returns NULL.

```c
enum pmempool_check_result pmempool_check_end(PMEMpoolcheck* ppc);
```

The **pmempool_check_end**() function finalizes the check and
releases all related resources. *ppc* is not a valid
pointer after calling **pmempool_check_end**(). It
returns *enum pmempool_check_result* summarizing the results
of the finalized check. **pmempool_check_end**() can
return one of the following values:

+ **PMEMPOOL_CHECK_RESULT_CONSISTENT** - the *pool* is consistent
+ **PMEMPOOL_CHECK_RESULT_NOT_CONSISTENT** - the *pool* is not consistent
+ **PMEMPOOL_CHECK_RESULT_REPAIRED** - the *pool* has issues but all repair
  steps completed successfully
+ **PMEMPOOL_CHECK_RESULT_CANNOT_REPAIR** - the *pool* has issues which
  can not be repaired
+ **PMEMPOOL_CHECK_RESULT_ERROR** - the *pool* has errors or the check
  encountered issues


# POOL SET SYNCHRONIZATION AND TRANSFORMATION #

Currently, the following operations are allowed only for **pmemobj** pools (see
**libpmemobj**(3)).


### POOL SET SYNC ###

```c
_UWFUNCR1(int, pmempool_sync, *poolset_file,_q_
	unsigned flags_e_, _q_ (EXPERIMENTAL)_e_)
```

The _UW(pmempool_sync) function synchronizes data between replicas within
a pool set.

_UW(pmempool_sync) accepts two arguments:

* *poolset_file* - a path to a pool set file,

* *flags* - a combination of flags (ORed) which modify the way of
synchronization.

>NOTE: Only the pool set file used to create the pool should be used
for syncing the pool.

The following flags are available:

* **PMEMPOOL_DRY_RUN** - do not apply changes, only check for viability of
synchronization.

The _UW(pmempool_sync) function checks if metadata of all replicas in a pool set
are consistent, i.e. all parts are healthy, and if any of them is not,
the corrupted or missing parts are recreated and filled with data from one of
the healthy replicas.

The function returns either 0 on success or -1 in case of error
with errno set accordingly.

>NOTE: The _UW(pmempool_sync) API is experimental and it may change in future
versions of the library.

### POOL SET TRANSFORM ###

```c
_UWFUNCR12(int, pmempool_transform, *poolset_file_src,
	*poolset_file_dst, unsigned flags, _q_ (EXPERIMENTAL)_e_)
```

The _UW(pmempool_transform) function modifies the internal structure of a pool set.
It supports the following operations:

* adding one or more replicas,

* removing one or more replicas,

* reordering of replicas.


_UW(pmempool_transform) accepts three arguments:

* *poolset_file_src* - a path to a pool set file which defines the source
pool set to be changed,

* *poolset_file_dst* - a path to a pool set file which defines the target
structure of the pool set,

* *flags* - a combination of flags (ORed) which modify the way of
synchronization.

The following flags are available:

* **PMEMPOOL_DRY_RUN** - do not apply changes, only check for viability of
synchronization.

When adding or deleting replicas, the two pool set files can differ only in the
definitions of replicas which are to be added or deleted. One cannot add and
remove replicas in the same step. Only one of these operations can be performed
at a time. Reordering replicas can be combined with any of them.
Also, to add a replica it is necessary for its effective size to match or exceed
the pool size. Otherwise the whole operation fails and no changes are applied.
Effective size of a replica is the sum of sizes of all its part files decreased
by 4096 bytes per each part file. The 4096 bytes of each part file is
utilized for storing internal metadata of the pool part files.

The function returns either 0 on success or -1 in case of error
with *errno* set accordingly.

>NOTE: The _UW(pmempool_transform) API is experimental and it may change in future
versions of the library.


# POOL SET MANAGEMENT FUNCTIONS: #

### Removing pool ###

```c
_UWFUNCR1(int, pmempool_rm, *path, int flags)
```

The _UW(pmempool_rm) function removes the pool pointed to by *path*. The *path* can
point to a regular file, device dax or pool set file. If *path* is a pool
set file, _UW(pmempool_rm) will remove all part files from local replicas
using **unlink**(3), _WINUX(,_q_and all remote replicas using **rpmem_remove**()
(see **librpmem**(3)),_e_) before removing the pool set file itself.

The *flags* argument determines the behavior of _UW(pmempool_rm).
It is either 0 or the bitwise OR of one or more of the following flags:

+ **PMEMPOOL_RM_FORCE**
Ignore all errors when removing part files from local replicas or remote
replica.

+ **PMEMPOOL_RM_POOLSET_LOCAL**
Remove also local pool set file.

+ **PMEMPOOL_RM_POOLSET_REMOTE**
Remove also remote pool set file.


# CAVEATS #

**libpmempool** relies on the library destructor being called from the main
thread. For this reason, all functions that might trigger destruction (e.g.
**dlclose**()) should be called in the main thread. Otherwise some of the
resources associated with that thread might not be cleaned up properly.


# LIBRARY API VERSIONING #

This section describes how the library API is versioned, allowing
applications to work with an evolving API.

```c
_UWFUNC(pmempool_check_version, _q_
	unsigned major_required,
	unsigned minor_required_e_)
```

The _UW(pmempool_check_version) function is used to see if
the installed **libpmempool** supports the version of the
library API required by an application. The easiest way to
do this for the application is to supply the compile-time
version information, supplied by defines in **\<libpmempool.h\>**, like this:

```c
reason = _U(pmempool_check_version)(PMEMPOOL_MAJOR_VERSION,
                                PMEMPOOL_MINOR_VERSION);
if (reason != NULL) {
	/* version check failed, reason string tells you why */
}
```

Any mismatch in the major version number is considered a failure, but a
library with a newer minor version number will pass this
check since increasing minor versions imply backwards compatibility.

An application can also check specifically for the existence of an
interface by checking for the version where that interface
was introduced. These versions are documented in this man
page as follows: unless otherwise specified, all interfaces
described here are available in version 1.0 of the library.
Interfaces added after version 1.0 will contain the text
*introduced in version x.y* in the section of this manual
describing the feature.

When the version check performed by _UW(pmempool_check_version)
is successful, the return value is NULL. Otherwise the
return value is a static string describing the reason for
failing the version check. The string returned by
_UW(pmempool_check_version) must not be modified or freed.


# DEBUGGING AND ERROR HANDLING #

Two versions of libpmempool are typically available on a development
system. The normal version, accessed when a program is
linked using the **-lpmempool** option, is optimized for
performance. That version skips checks that impact
performance and exceptionally logs any trace information or
performs any run-time assertions. If an error is detected
during the call to **libpmempool** function, an
application may retrieve an error message describing the
reason of failure using the following function:

```c
_UWFUNC(pmempool_errormsg, void)
```

The _UW(pmempool_errormsg) function returns a pointer to a
static buffer containing the last error message logged for
current thread. The error message may include description of
the corresponding error code (if *errno* was set), as returned
by **strerror**(3). The error message buffer is
thread-local; errors encountered in one thread do not affect
its value in other threads. The buffer is never cleared by
any library function; its content is significant only when
the return value of the immediately preceding call to
**libpmempool** function indicated an error, or if *errno*
was set. The application must not modify or free the error
message string, but it may be modified by subsequent calls
to other library functions.

A second version of **libpmempool**, accessed when a program uses
the libraries under _WINUX(**/nvml/src/x64/Debug**,**/usr/lib/nvml_debug**), contains
run-time assertions and trace points. The typical way to
access the debug version is to set the environment variable
**LD_LIBRARY_PATH** to _WINUX(**/nvml/src/x64/Debug** or other location,
**/usr/lib/nvml_debug** or **/usr/lib64/nvml_debug**), depending on where the debug
libraries are installed on the system.
The trace points in
the debug version of the library are enabled using the
environment variable **PMEMPOOL_LOG_LEVEL**, which can be
set to the following values:

+ **0** - This is the default level when **PMEMPOOL_LOG_LEVEL** is not set.
No log messages are emitted at this level.

+ **1** - Additional details on any errors detected are logged (in addition to
returning the *errno*-based errors as usual). The same information may be
retrieved using _UW(pmempool_errormsg).

+ **2** - A trace of basic operations is logged.

+ **3** - This level enables a very verbose amount of function call tracing in
the library.

+ **4** - This level enables voluminous and fairly obscure tracing information
that is likely only useful to the libpmempool developers.

The environment variable **PMEMPOOL_LOG_FILE** specifies a file name
where all logging information should be written. If the last
character in the name is "-", the PID of the
current process will be appended to the file name when the
log file is created. If **PMEMPOOL_LOG_FILE** is not set,
the logging output goes to stderr.

Setting the environment variable **PMEMPOOL_LOG_FILE** has no effect
on the non-debug version of **libpmempool**.


# EXAMPLE #

The following example illustrates how the **libpmempool** API is used.
The program detects the type and checks consistency of given pool.
If there are any issues detected, the pool is automatically repaired.

```c
#include <stddef.h>_WINUX(,_q_
#include <unistd.h>_e_)
#include <stdlib.h>
#include <stdio.h>
#include <libpmempool.h>

#define PATH "./pmem-fs/myfile"
#define CHECK_FLAGS (PMEMPOOL_CHECK_FORMAT_STR|PMEMPOOL_CHECK_REPAIR|\
                     PMEMPOOL_CHECK_VERBOSE)

int
main(int argc, char *argv[])
{
	PMEMpoolcheck *ppc;
	struct _U(pmempool_check_status) *status;
	enum pmempool_check_result ret;

	/* arguments for check */
	struct _U(pmempool_check_args) args = {
		.path		= PATH,
		.backup_path	= NULL,
		.pool_type	= PMEMPOOL_POOL_TYPE_DETECT,
		.flags		= CHECK_FLAGS
	};

	/* initialize check context */
	if ((ppc = _U(pmempool_check_init)(&args, sizeof(args))) == NULL) {
		perror("_U(pmempool_check_init)");
		exit(EXIT_FAILURE);
	}

	/* perform check and repair, answer 'yes' for each question */
	while ((status = _U(pmempool_check)(ppc)) != NULL) {
		switch (status->type) {
		case PMEMPOOL_CHECK_MSG_TYPE_ERROR:
			printf("%s\n", status->str.msg);
			break;
		case PMEMPOOL_CHECK_MSG_TYPE_INFO:
			printf("%s\n", status->str.msg);
			break;
		case PMEMPOOL_CHECK_MSG_TYPE_QUESTION:
			printf("%s\n", status->str.msg);
			status->str.answer = "yes";
			break;
		default:
			pmempool_check_end(ppc);
			exit(EXIT_FAILURE);
		}
	}

	/* finalize the check and get the result */
	ret = pmempool_check_end(ppc);
	switch (ret) {
		case PMEMPOOL_CHECK_RESULT_CONSISTENT:
		case PMEMPOOL_CHECK_RESULT_REPAIRED:
			return 0;
		default:
			return 1;
	}
}
```

See <http://pmem.io/nvml/libpmempool> for more examples using the
**libpmempool** API.


# ACKNOWLEDGEMENTS #

**libpmempool** builds on the persistent memory programming model
recommended by the SNIA NVM Programming Technical Work Group:
<http://snia.org/nvmp>


# SEE ALSO #

**mmap**(2), **munmap**(2), **msync**(2), **strerror**(3),
**libpmemobj**(3), **libpmemblk**(3), **libpmemlog**(3), **libpmem**(3)
and **<http://pmem.io>**


src/libpmempool/libpmempool.c
/*
 * Copyright 2016-2017, Intel Corporation
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *     * Redistributions of source code must retain the above copyright
 *       notice, this list of conditions and the following disclaimer.
 *
 *     * Redistributions in binary form must reproduce the above copyright
 *       notice, this list of conditions and the following disclaimer in
 *       the documentation and/or other materials provided with the
 *       distribution.
 *
 *     * Neither the name of the copyright holder nor the names of its
 *       contributors may be used to endorse or promote products derived
 *       from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/*
 * libpmempool.c -- entry points for libpmempool
 */

#include <stdlib.h>
#include <stdint.h>
#include <errno.h>
#include <sys/param.h>

#include "pmemcommon.h"
#include "libpmempool.h"
#include "pmempool.h"
#include "pool.h"
#include "check.h"

#ifdef USE_RPMEM
#include "rpmem_common.h"
#include "rpmem_util.h"
#endif

#ifdef _WIN32
#define ANSWER_BUFFSIZE 256
#endif

/*
 * libpmempool_init -- load-time initialization for libpmempool
 *
 * Called automatically by the run-time loader.
 */
__attribute__((constructor))
static void
libpmempool_init(void)
{
	common_init(PMEMPOOL_LOG_PREFIX, PMEMPOOL_LOG_LEVEL_VAR,
		PMEMPOOL_LOG_FILE_VAR, PMEMPOOL_MAJOR_VERSION,
		PMEMPOOL_MINOR_VERSION);
	LOG(3, NULL);
#ifdef USE_RPMEM
	if (util_remote_init()) {
		LOG(4, "Duplicate util_remote_init()");
	}
	rpmem_util_cmds_init();
#endif
}

/*
 * libpmempool_fini -- libpmempool cleanup routine
 *
 * Called automatically when the process terminates.
 */
__attribute__((destructor))
static void
libpmempool_fini(void)
{
	LOG(3, NULL);
#ifdef USE_RPMEM
	util_remote_unload();
	util_remote_fini();
	rpmem_util_cmds_fini();
#endif
	common_fini();
}

/*
 * pmempool_check_versionU -- see if library meets application version
 *	requirements
 */
#ifndef _WIN32
static inline
#endif
const char *
pmempool_check_versionU(unsigned major_required, unsigned minor_required)
{
	LOG(3, "major_required %u minor_required %u",
			major_required, minor_required);

	if (major_required != PMEMPOOL_MAJOR_VERSION) {
		ERR("libpmempool major version mismatch (need %u, found %u)",
			major_required, PMEMPOOL_MAJOR_VERSION);
		return out_get_errormsg();
	}

	if (minor_required > PMEMPOOL_MINOR_VERSION) {
		ERR("libpmempool minor version mismatch (need %u, found %u)",
			minor_required, PMEMPOOL_MINOR_VERSION);
		return out_get_errormsg();
	}

	return NULL;
}

#ifndef _WIN32
/*
 * pmempool_check_version -- see if lib meets application version requirements
 */
const char *
pmempool_check_version(unsigned major_required, unsigned minor_required)
{
	return pmempool_check_versionU(major_required, minor_required);
}
#else
/*
 * pmempool_check_versionW -- see if library meets application version
 *	requirements as widechar
 */
const wchar_t *
pmempool_check_versionW(unsigned major_required, unsigned minor_required)
{
	if (pmempool_check_versionU(major_required, minor_required) != NULL)
		return out_get_errormsgW();
	else
		return NULL;
}
#endif

/*
 * pmempool_errormsgU -- return last error message
 */
#ifndef _WIN32
static inline
#endif
const char *
pmempool_errormsgU(void)
{
	return out_get_errormsg();
}

#ifndef _WIN32
/*
 * pmempool_errormsg -- return last error message
 */
const char *
pmempool_errormsg(void)
{
	return pmempool_errormsgU();
}
#else
/*
 * pmempool_errormsgW -- return last error message as widechar
 */
const wchar_t *
pmempool_errormsgW(void)
{
	return out_get_errormsgW();
}
#endif

/*
 * pmempool_ppc_set_default -- (internal) set default values of check context
 */
static void
pmempool_ppc_set_default(PMEMpoolcheck *ppc)
{
	const PMEMpoolcheck ppc_default = {
		.args		= {
			.path		= NULL,
			.backup_path	= NULL,
			.pool_type	= PMEMPOOL_POOL_TYPE_DETECT,
			.flags		= 0
		},
		.data		= NULL,
		.pool		= NULL,
		.result		= CHECK_RESULT_CONSISTENT,
	};
	*ppc = ppc_default;
}

/*
 * pmempool_check_initU -- initialize check context
 */
#ifndef _WIN32
static inline
#endif
PMEMpoolcheck *
pmempool_check_initU(struct pmempool_check_argsU *args, size_t args_size)
{
	LOG(3, "path %s backup_path %s pool_type %u flags %x", args->path,
		args->backup_path, args->pool_type, args->flags);

	/*
	 * Currently one size of args structure is supported. The version of the
	 * pmempool_check_args structure can be distinguished based on provided
	 * args_size.
	 */
	if (args_size < sizeof(struct pmempool_check_args)) {
		ERR("provided args_size is not supported");
		errno = EINVAL;
		return NULL;
	}

	/*
	 * Dry run does not allow to made changes possibly performed during
	 * repair. Advanced allow to perform more complex repairs. Questions
	 * are ask only if repairs are made. So dry run, advanced and always_yes
	 * can be set only if repair is set.
	 */
	if (util_flag_isclr(args->flags, PMEMPOOL_CHECK_REPAIR) &&
			util_flag_isset(args->flags, PMEMPOOL_CHECK_DRY_RUN |
			PMEMPOOL_CHECK_ADVANCED | PMEMPOOL_CHECK_ALWAYS_YES)) {
		ERR("dry_run, advanced and always_yes are applicable only if "
			"repair is set");
		errno = EINVAL;
		return NULL;
	}

	/*
	 * dry run does not modify anything so performing backup is redundant
	 */
	if (util_flag_isset(args->flags, PMEMPOOL_CHECK_DRY_RUN) &&
			args->backup_path != NULL) {
		ERR("dry run does not allow to perform backup");
		errno = EINVAL;
		return NULL;
	}

	/*
	 * libpmempool uses str format of communication so it must be set
	 */
	if (util_flag_isclr(args->flags, PMEMPOOL_CHECK_FORMAT_STR)) {
		ERR("PMEMPOOL_CHECK_FORMAT_STR flag must be set");
		errno = EINVAL;
		return NULL;
	}

	PMEMpoolcheck *ppc = malloc(sizeof(*ppc));
	if (ppc == NULL) {
		ERR("!malloc");
		return NULL;
	}

	pmempool_ppc_set_default(ppc);
	memcpy(&ppc->args, args, sizeof(ppc->args));
	ppc->path = strdup(args->path);
	if (!ppc->path) {
		ERR("!strdup");
		goto error_path_malloc;
	}
	ppc->args.path = ppc->path;

	if (args->backup_path != NULL) {
		ppc->backup_path = strdup(args->backup_path);
		if (!ppc->backup_path) {
			ERR("!strdup");
			goto error_backup_path_malloc;
		}
		ppc->args.backup_path = ppc->backup_path;
	}

	if (check_init(ppc) != 0)
		goto error_check_init;

	return ppc;

error_check_init:
	/* in case errno not set by any of the used functions set its value */
	if (errno == 0)
		errno = EINVAL;

	free(ppc->backup_path);
error_backup_path_malloc:
	free(ppc->path);
error_path_malloc:
	free(ppc);
	return NULL;
}

#ifndef _WIN32
/*
 * pmempool_check_init -- initialize check context
 */
PMEMpoolcheck *
pmempool_check_init(struct pmempool_check_args *args, size_t args_size)
{
	return pmempool_check_initU(args, args_size);
}
#else
/*
 * pmempool_check_initW -- initialize check context as widechar
 */
PMEMpoolcheck *
pmempool_check_initW(struct pmempool_check_argsW *args, size_t args_size)
{
	char *upath = util_toUTF8(args->path);
	if (upath == NULL)
		return NULL;
	char *ubackup_path = NULL;
	if (args->backup_path != NULL) {
		ubackup_path = util_toUTF8(args->backup_path);
		if (ubackup_path == NULL) {
			util_free_UTF8(upath);
			return NULL;
		}
	}

	struct pmempool_check_argsU uargs = {
		.path = upath,
		.backup_path = ubackup_path,
		.pool_type = args->pool_type,
		.flags = args->flags
	};

	PMEMpoolcheck *ret = pmempool_check_initU(&uargs, args_size);

	util_free_UTF8(ubackup_path);
	util_free_UTF8(upath);
	return ret;
}
#endif

/*
 * pmempool_checkU -- continue check till produce status to consume for caller
 */
#ifndef _WIN32
static inline
#endif
struct pmempool_check_statusU *
pmempool_checkU(PMEMpoolcheck *ppc)
{
	LOG(3, NULL);
	ASSERTne(ppc, NULL);

	struct check_status *result;
	do {
		result = check_step(ppc);

		if (check_is_end(ppc->data) && result == NULL)
			return NULL;
	} while (result == NULL);

	return check_status_get(result);
}

#ifndef _WIN32
/*
 * pmempool_check -- continue check till produce status to consume for caller
 */
struct pmempool_check_status *
pmempool_check(PMEMpoolcheck *ppc)
{
	return pmempool_checkU(ppc);
}
#else
/*
 * pmempool_checkW -- continue check till produce status to consume for caller
 */
struct pmempool_check_statusW *
pmempool_checkW(PMEMpoolcheck *ppc)
{
	LOG(3, NULL);
	ASSERTne(ppc, NULL);

	/* check the cache and convert msg and answer */
	char buf[ANSWER_BUFFSIZE];
	memset(buf, 0, ANSWER_BUFFSIZE);
	convert_status_cache(ppc, buf, ANSWER_BUFFSIZE);

	struct check_status *uresult;
	do {
		uresult = check_step(ppc);

		if (check_is_end(ppc->data) && uresult == NULL)
			return NULL;
	} while (uresult == NULL);

	struct pmempool_check_statusU *uret_res = check_status_get(uresult);
	const wchar_t *wmsg = util_toUTF16(uret_res->str.msg);
	if (wmsg == NULL)
		FATAL("!malloc");

	struct pmempool_check_statusW *wret_res =
		(struct pmempool_check_statusW *)uret_res;
	/* pointer to old message is freed in next check step */
	wret_res->str.msg = wmsg;
	return wret_res;
}
#endif

/*
 * pmempool_check_end -- end check and release check context
 */
enum pmempool_check_result
pmempool_check_end(PMEMpoolcheck *ppc)
{
	LOG(3, NULL);
	enum check_result result = ppc->result;

	check_fini(ppc);
	free(ppc->path);
	free(ppc->backup_path);
	free(ppc);

	switch (result) {
		case CHECK_RESULT_CONSISTENT:
			return PMEMPOOL_CHECK_RESULT_CONSISTENT;

		case CHECK_RESULT_NOT_CONSISTENT:
			return PMEMPOOL_CHECK_RESULT_NOT_CONSISTENT;

		case CHECK_RESULT_REPAIRED:
			return PMEMPOOL_CHECK_RESULT_REPAIRED;

		case CHECK_RESULT_CANNOT_REPAIR:
			return PMEMPOOL_CHECK_RESULT_CANNOT_REPAIR;

		default:
			return PMEMPOOL_CHECK_RESULT_ERROR;
	}
}
